RecoHGCal/TICL/plugins/FilteredLayerClustersProducer.cc

```diff
@@ -74,8 +74,8 @@ void FilteredLayerClustersProducer::produce(edm::Event& evt, const edm::EventSet
   edm::Handle<std::vector<reco::CaloCluster>> clusterHandle;
   edm::Handle<std::vector<float>> inputClustersMaskHandle;
   auto availableLayerClusters = std::make_unique<ticl::HgcalClusterFilterMask>();
-  evt.getByToken(clusters_token_, clusterHandle);
-  evt.getByToken(clustersMask_token_, inputClustersMaskHandle);
+  clusterHandle = evt.getHandle(clusters_token_);
+  inputClustersMaskHandle = evt.getHandle(clustersMask_token_);
```

Comment:
It would be nice to be able to go directly to  e.g.
```suggestion
  auto clusterHandle = evt.getHandle(clusters_token_);
  auto inputClustersMaskHandle = evt.getHandle(clustersMask_token_);
```
and remove the lines 74 and 75, but what I understood from @gartung this would be difficult-to-impossible to do that in clang-tidy in general.

Reply:
> It would be nice to be able to go directly to e.g.

so, then what is the plan? Are we going to automatically migrate to getHandle and then expect someone elides the use of handles?

Reply:
it may be easier to leave the old interface unless someone checks the code manually

Reply:
For the specific case that a `edm::Handle` variable is declared in the very same scope as the `Event::getHandle()` is being called, it could be possible to have some automatic post processing to combine them into one line.

Any manual action could also be left for a later edit of that file (e.g. when someone becomes irritated-enough by the pattern produced by the automatic conversion; at least the automated migration would it more clear where code could be simplified later, even by going to the `Event::get()`).

Reply:
> Any manual action could also be left for a later edit of that file (e.g. when someone becomes irritated-enough by the pattern produced by the automatic conversion; at least the automated migration would it more clear where code could be simplified later, even by going to the `Event::get()`).

but isn't the point of not migrating in this case to be that there is a tool like a static analyzer that would point out the old/bad cases?
After the migration there will be no clear way to highlight problematic cases.

Reply:
If it is deemed that a manual edits would be needed anyway in many cases, then a static analyzer / clang-tidy check at code-checks or PR test phase would likely work better than fully automatic migration.

With that approach my concern is that migrating the ~7.5k calls would take years to complete.

Maybe a topic for a future core software (or release planning) meeting?

Reply:
how much of this can be automated?
E.g. is it possible to replace `edm::Handle<Type> foo; e.getByToken(tok, foo);` with `Type const* foo = nullptr; foo = &e.get(tok);` ( perhaps contracted if the scope is the same, maybe restrict to scopes without `foo.` ), then try to compile and use it if it compiles?

I'm guessing that this will cover about a half of the cases (including this file).

Reply:
BTW, if we are gaining in execution cost by using `evt.getHandle(`, then it should be OK to migrate automatically but to add smth like `//auto-migrated: check if product can be used directly with .get`
and then report all of these comments in the static analysis summary

Reply:
There is no performance difference between `Event::getByToken()` and `Event:getHandle()` in practice (in principle latter makes two assignments and one `std::shared_ptr::reset()` less). `Event::get()` would give a (small) performance gain.

I suggest we discuss more on the possibilities for further automation in the core meeting tomorrow.